lib/charms/identity_platform_login_ui_operator/v0/login_ui_endpoints.py

```diff
@@ -0,0 +1,165 @@
+```shell
+cd some-charm
+charmcraft fetch-lib charms.identity_platform_login_ui.v0.login_ui_endpoints
+```
+To use the library from the requirer side:
+In the `metadata.yaml` of the charm, add the following:
+```yaml
+requires:
+  ui-endpoints-info:
+    interface: login_ui_endpoints
+    limit: 1
+```
+Then, to initialise the library:
+```python
+from charms.identity_platform_login_ui.v0.login_ui_endpoints import (
+    LoginUIEndpointsRelationError,
+    LoginUIEndpointsRequirer,
+)
+Class SomeCharm(CharmBase):
+    def __init__(self, *args):
+        self.login_ui_endpoints_relation = LoginUIEndpointsRequirer(self)
+        self.framework.observe(self.on.some_event_emitted, self.some_event_function)
+    def some_event_function():
+        # fetch the relation info
+        try:
+            login_ui_endpoints = self.login_ui_endpoints_relation.get_login_ui_endpoints()
+        except LoginUIEndpointsRelationError as error:
+            ...
+```
+"""
+
+import logging
+from typing import Dict, Optional
+
+from ops.charm import CharmBase, RelationCreatedEvent
+from ops.framework import EventBase, EventSource, Object, ObjectEvents
+
+# The unique Charmhub library identifier, never change it
+LIBID = "460ab09e6b874d1c891b67f83586c9a7"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+RELATION_NAME = "ui-endpoint-info"
+INTERFACE_NAME = "login_ui_endpoints"
+logger = logging.getLogger(__name__)
+
+RELATION_KEYS = [
+    "consent_url",
+    "error_url",
+    "index_url",
+    "login_url",
+    "oidc_error_url",
+    "registration_url",
+    "default_url",
+]
+
+
+class LoginUIEndpointsRelationReadyEvent(EventBase):
+    """Event to notify the charm that the relation is ready."""
+
+
+class LoginUIEndpointsProviderEvents(ObjectEvents):
+    """Event descriptor for events raised by `LoginUIEndpointsProvider`."""
+
+    ready = EventSource(LoginUIEndpointsRelationReadyEvent)
+
+
+class LoginUIEndpointsProvider(Object):
+    """Provider side of the endpoint-info relation."""
+
+    on = LoginUIEndpointsProviderEvents()
+
+    def __init__(self, charm: CharmBase, relation_name: str = RELATION_NAME):
+        super().__init__(charm, relation_name)
+
+        self._charm = charm
+        self._relation_name = relation_name
+
+        events = self._charm.on[relation_name]
+        self.framework.observe(
+            events.relation_created, self._on_provider_endpoints_relation_created
+        )
+
+    def _on_provider_endpoints_relation_created(self, event: RelationCreatedEvent) -> None:
+        self.on.ready.emit()
+
+    def send_endpoints_relation_data(self, endpoint: str) -> None:
+        """Updates relation with endpoint info."""
+        if not self._charm.unit.is_leader():
+            return
```

Comment:
Since this is in a library, we may want to raise here (instead of returning) so that the user knows has feedback that they shouldn't call this method from non-leader units. And in the charm you will want to make sure that it's only the leader that would call this method.

Reply:
The function now raises an error, which is handled in charm.py